plugins/via-usbhub/fu-via-usbhub-device.c

```diff
@@ -0,0 +1,1247 @@
+		self->kind = FU_VIA_USBHUB_DEVICE_KIND_VL811;
+	} else if ((b811P812 & ((1 << 5) | (1 << 4))) == 0) {
+		if (chipver == 0x10)
+			self->kind = FU_VIA_USBHUB_DEVICE_KIND_VL811PB0;
+		else
+			self->kind = FU_VIA_USBHUB_DEVICE_KIND_VL811PB3;
+	} else if ((b811P812 & ((1 << 5) | (1 << 4))) == (1 << 4)) {
+		self->kind = FU_VIA_USBHUB_DEVICE_KIND_VL812Q4S;
+	} else if ((b811P812 & ((1 << 5) | (1 << 4))) == ((1 << 5) | (1 << 4))) {
+		if (chipver == 0x10)
+			self->kind = FU_VIA_USBHUB_DEVICE_KIND_VL812B0;
+		else
+			self->kind = FU_VIA_USBHUB_DEVICE_KIND_VL812B3;
+	} else {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "hardware is not supported");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static GBytes *
+fu_via_usbhub_device_dump_firmware (FuViaUsbhubDevice *self, gsize bufsz, GError **error)
+{
+	g_autofree guint8 *buf = g_malloc0 (bufsz);
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* get data from hardware */
+	chunks = fu_chunk_array_new (buf, bufsz, 0x0, 0x0, FU_VIA_USBHUB_TXSIZE);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index (chunks, i);
+		if (!fu_via_usbhub_device_spi_read_data (self,
+							 chk->address,
+							 (guint8 *) chk->data,
+							 chk->data_sz,
+							 error)) {
+			g_prefix_error (error, "SPI data verify failed @0x%x: ", chk->address);
+			return FALSE;
+		}
+		fu_device_set_progress_full (FU_DEVICE (self),
+					     (gsize) i, (gsize) chunks->len);
+	}
+	return g_bytes_new_take (g_steal_pointer (&buf), bufsz);
+}
+
+static gboolean
+fu_via_usbhub_device_setup (FuDevice *device, GError **error)
+{
+	FuViaUsbhubDevice *self = FU_VIA_USBHUB_DEVICE (device);
+	const guint8 *buf;
+	guint32 flash_id = 0x0;
+	guint8 buf_tmp[FU_VIA_USBHUB_TXSIZE] = { 0x0 };
+	g_autoptr(GError) error_tmp = NULL;
+	g_autoptr(GBytes) fw = NULL;
+
+	/* try to read a block of data which will fail for 813-type devices */
+	if (fu_device_has_custom_flag (device, "needs-unlock-legacy813") &&
+	    !fu_via_usbhub_device_spi_read_data (self, 0x0, buf_tmp, sizeof(buf_tmp), &error_tmp)) {
+		g_warning ("failed to read, trying to unlock 813: %s", error_tmp->message);
+		if (!fu_via_usbhub_device_vdr_unlock_813 (self, error))
+			return FALSE;
+		if (!fu_via_usbhub_device_spi_read_data (self, 0x0, buf_tmp, sizeof(buf_tmp), error)) {
+			g_prefix_error (error, "813 unlock fail: ");
+			return FALSE;
+		}
+		g_debug ("813 unlock OK");
+		/* VL813 & VL210 have same PID (0x0813), and only VL813 can reply */
+		self->kind = FU_VIA_USBHUB_DEVICE_KIND_VL813;
+	} else {
+		if (!fu_via_usbhub_device_guess_kind (self, error))
+			return FALSE;
+	}
+
+	/* get the flash chip attached */
+	if (!fu_via_usbhub_device_spi_read_flash_id (self, &flash_id, error)) {
+		g_prefix_error (error, "failed to read SPI chip ID: ");
+		return FALSE;
+	}
+	if (flash_id != 0x0) {
+		g_autofree gchar *instance_id = NULL;
+		instance_id = g_strdup_printf ("SPI\\FLASH_%04X", flash_id);
```

Comment:
Is this generic "SPI" domain because you are thinking that SPI flash configurations might carry over to another plugin some day?  I think it should be VIA domain.

Reply:
Yup, that was my thinking, and I think you're probably right. What about `VIA_SPI\FLASH_FOO`?

Reply:
`VIA_USBHUB\SPI_C220` works better

Reply:
Sounds good.